rpc: TestHeartbeatHealthTransport failed

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4566826&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4566826&tab=artifacts#/) on master @ [231fd420c61e39b6ad08c982c496d82cf1910bc5](https://github.com/cockroachdb/cockroach/commits/231fd420c61e39b6ad08c982c496d82cf1910bc5):


```
=== RUN   TestHeartbeatHealthTransport
    context_test.go:757: rpc error: code = Unknown desc = client cluster ID "5745c6bc-64a8-40f4-9d60-a9dc253c27e8" doesn't match server cluster ID "f8521426-68e9-429d-abd3-99406a9db4de"
--- FAIL: TestHeartbeatHealthTransport (0.05s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=bazel,gss,deadlock
</p>
</details>
/cc @cockroachdb/server
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-13752

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8089750?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8089750?buildTab=artifacts#/) on master @ [c4bde8b72cdd4016845ae70ef5162b3f11fab1fb](https://github.com/cockroachdb/cockroach/commits/c4bde8b72cdd4016845ae70ef5162b3f11fab1fb):


```
=== RUN   TestHeartbeatHealthTransport
I221224 08:46:10.259654 4031 2@rpc/context.go:2297  [rnode=1,raddr=127.0.0.1:40387,class=default,rpc] 12  connection is now ready
E221224 08:46:10.269621 4031 2@rpc/context.go:2137  [rnode=1,raddr=127.0.0.1:40387,class=default,rpc] 13  closing connection after: grpc: error reading from server: read tcp 127.0.0.1:51202->127.0.0.1:40387: read: connection reset by peer [code 14/Unavailable]
I221224 08:46:10.286447 6151 2@rpc/context.go:2297  [rnode=1,raddr=127.0.0.1:40387,class=default,rpc] 14  connection is now ready
E221224 08:46:10.287262 6151 2@rpc/context.go:2137  [rnode=1,raddr=127.0.0.1:40387,class=default,rpc] 15  closing connection after: grpc: error reading from server: read tcp 127.0.0.1:51208->127.0.0.1:40387: read: connection reset by peer [code 14/Unavailable]
E221224 08:46:10.297873 7266 2@rpc/context.go:2137  [rnode=1,raddr=127.0.0.1:40387,class=default,rpc] 16  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:40387: connect: connection refused" [code 14/Unavailable]
    context_test.go:1268: initial connection heartbeat failed: grpc: client cluster ID "497df59c-7d6c-417b-b918-46ea813398ca" doesn't match server cluster ID "3ffe7241-9004-4264-b4cc-530505bff341" [code 2/Unknown]
--- FAIL: TestHeartbeatHealthTransport (0.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8149790?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8149790?buildTab=artifacts#/) on master @ [0725273ac7f789ba8ed78aacaf73cc953ca47fe8](https://github.com/cockroachdb/cockroach/commits/0725273ac7f789ba8ed78aacaf73cc953ca47fe8):


```
=== RUN   TestHeartbeatHealthTransport
I221231 09:36:30.224178 447 2@rpc/context.go:2297  [rnode=1,raddr=127.0.0.1:44759,class=default,rpc] 11  connection is now ready
E221231 09:36:30.228723 447 2@rpc/context.go:2137  [rnode=1,raddr=127.0.0.1:44759,class=default,rpc] 12  closing connection after: grpc: error reading from server: read tcp 127.0.0.1:34438->127.0.0.1:44759: read: connection reset by peer [code 14/Unavailable]
I221231 09:36:30.244513 520 2@rpc/context.go:2297  [rnode=1,raddr=127.0.0.1:44759,class=default,rpc] 13  connection is now ready
E221231 09:36:30.244910 520 2@rpc/context.go:2137  [rnode=1,raddr=127.0.0.1:44759,class=default,rpc] 14  closing connection after: grpc: connection error: desc = "transport: Error while dialing connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
E221231 09:36:30.245490 73 2@rpc/context.go:2137  [rnode=1,raddr=127.0.0.1:44759,class=default,rpc] 15  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:44759: connect: connection refused" [code 14/Unavailable]
    context_test.go:1268: initial connection heartbeat failed: grpc: client cluster ID "c9778921-89db-409c-9252-19dbb9cd53b0" doesn't match server cluster ID "3f586a6f-2aaa-4c0f-a8bf-127a494bf15c" [code 2/Unknown]
--- FAIL: TestHeartbeatHealthTransport (0.06s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8581583?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8581583?buildTab=artifacts#/) on master @ [5fbcd8a8deac0205c7df38e340c1eb9692854383](https://github.com/cockroachdb/cockroach/commits/5fbcd8a8deac0205c7df38e340c1eb9692854383):


```
=== RUN   TestHeartbeatHealthTransport
I230204 08:19:12.321807 4522 2@rpc/context.go:2412  [T1,rnode=1,raddr=127.0.0.1:40627,class=default,rpc] 13  connection is now ready
E230204 08:19:12.322123 4522 2@rpc/context.go:2250  [T1,rnode=1,raddr=127.0.0.1:40627,class=default,rpc] 14  closing connection after: grpc: error reading from server: EOF [code 14/Unavailable]
I230204 08:19:12.338996 4553 2@rpc/context.go:2412  [T1,rnode=1,raddr=127.0.0.1:40627,class=default,rpc] 15  connection is now ready
E230204 08:19:12.339331 4553 2@rpc/context.go:2250  [T1,rnode=1,raddr=127.0.0.1:40627,class=default,rpc] 16  closing connection after: grpc: error reading from server: EOF [code 14/Unavailable]
E230204 08:19:12.339928 8037 2@rpc/context.go:2250  [T1,rnode=1,raddr=127.0.0.1:40627,class=default,rpc] 17  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:40627: connect: connection refused" [code 14/Unavailable]
I230204 08:19:12.367738 12477 2@rpc/context.go:2412  [T1,rnode=1,raddr=127.0.0.1:40627,class=default,rpc] 18  connection is now ready
    context_test.go:1282: <nil>
--- FAIL: TestHeartbeatHealthTransport (0.07s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8643641?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8643641?buildTab=artifacts#/) on master @ [09188370d82e163ff1d44c62fe611104502c548d](https://github.com/cockroachdb/cockroach/commits/09188370d82e163ff1d44c62fe611104502c548d):


```
=== RUN   TestHeartbeatHealthTransport
I230209 08:26:23.753269 4168 2@rpc/context.go:2472  [T1,rnode=1,raddr=127.0.0.1:34149,class=default,rpc] 13  connection is now ready
E230209 08:26:23.754031 4168 2@rpc/context.go:2310  [T1,rnode=1,raddr=127.0.0.1:34149,class=default,rpc] 14  closing connection after: grpc: error reading from server: read tcp 127.0.0.1:50126->127.0.0.1:34149: read: connection reset by peer [code 14/Unavailable]
I230209 08:26:23.771307 8533 2@rpc/context.go:2472  [T1,rnode=1,raddr=127.0.0.1:34149,class=default,rpc] 15  connection is now ready
E230209 08:26:23.771636 8533 2@rpc/context.go:2310  [T1,rnode=1,raddr=127.0.0.1:34149,class=default,rpc] 16  closing connection after: grpc: error reading from server: EOF [code 14/Unavailable]
E230209 08:26:23.772366 8607 2@rpc/context.go:2310  [T1,rnode=1,raddr=127.0.0.1:34149,class=default,rpc] 17  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:34149: connect: connection refused" [code 14/Unavailable]
    context_test.go:1282: initial connection heartbeat failed: grpc: client cluster ID "ab5a0d82-3ee0-4846-88f4-e6e14b17d486" doesn't match server cluster ID "b4b747b5-6d75-479f-bdce-42cc4a40a2eb" [code 2/Unknown]
--- FAIL: TestHeartbeatHealthTransport (0.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8868809?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8868809?buildTab=artifacts#/) on master @ [b0e5507f74c07e13cfda8cda8b9079b457a9f37d](https://github.com/cockroachdb/cockroach/commits/b0e5507f74c07e13cfda8cda8b9079b457a9f37d):


```
=== RUN   TestHeartbeatHealthTransport
I230228 08:47:11.235675 432 2@rpc/context.go:2473  [T1,rnode=1,raddr=127.0.0.1:44151,class=default,rpc] 13  connection is now ready
E230228 08:47:11.236001 432 2@rpc/context.go:2311  [T1,rnode=1,raddr=127.0.0.1:44151,class=default,rpc] 14  closing connection after: grpc: error reading from server: read tcp 127.0.0.1:42090->127.0.0.1:44151: read: connection reset by peer [code 14/Unavailable]
I230228 08:47:11.251944 477 2@rpc/context.go:2473  [T1,rnode=1,raddr=127.0.0.1:44151,class=default,rpc] 15  connection is now ready
E230228 08:47:11.252678 477 2@rpc/context.go:2311  [T1,rnode=1,raddr=127.0.0.1:44151,class=default,rpc] 16  closing connection after: grpc: error reading from server: EOF [code 14/Unavailable]
E230228 08:47:11.253432 480 2@rpc/context.go:2311  [T1,rnode=1,raddr=127.0.0.1:44151,class=default,rpc] 17  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: connection error: desc = "transport: Error while dialing dial tcp 127.0.0.1:44151: connect: connection refused" [code 14/Unavailable]
    context_test.go:1281: initial connection heartbeat failed: grpc: client cluster ID "375d134d-f59d-436d-96ec-02d9489742a9" doesn't match server cluster ID "456911e0-0407-43fd-96f4-f2925785904b" [code 2/Unknown]
--- FAIL: TestHeartbeatHealthTransport (0.05s)
```
<p>Parameters: <code>TAGS=bazel,gss</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9123591?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9123591?buildTab=artifacts#/) on master @ [0a72a49eb6e4ce5823b53d02125441c3da2e94f5](https://github.com/cockroachdb/cockroach/commits/0a72a49eb6e4ce5823b53d02125441c3da2e94f5):


```
E230318 08:19:05.437447 4687 2@rpc/context.go:2335  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 24  closing connection after: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
I230318 08:19:05.437535 4687 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 25  connection heartbeat loop ended with err: grpc: connection error: desc = "transport: error while dialing: connection interrupted (did the remote node shut down or are there networking issues?)" [code 14/Unavailable]
E230318 08:19:05.438245 8563 2@rpc/context.go:2335  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 26  unable to connect (is the peer up and reachable?): initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.438344 8563 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 27  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.438963 8742 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 28  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.439628 8874 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 29  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.440305 9022 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 30  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.440975 9175 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 31  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.441783 9310 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 32  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.442441 9458 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 33  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.442984 9582 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 34  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.443478 9736 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 35  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.444060 9327 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 36  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.444554 10005 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 37  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.446156 10125 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 38  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.447065 10308 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 39  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.448051 10382 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 40  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.448654 10505 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 41  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.449372 10636 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 42  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.449963 10790 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 43  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.455660 10862 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 44  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.456452 12214 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 45  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.458916 12345 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 46  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.459686 12890 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 47  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.460414 13023 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 48  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.461258 13204 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 49  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.461908 13371 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 50  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.462583 13485 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 51  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.463271 13650 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 52  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.463865 13768 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 53  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.464428 13707 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 54  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.466341 13988 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 55  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.467289 14280 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 56  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.468086 14406 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 57  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.469380 14501 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 58  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.470048 14706 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 59  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.470683 14847 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 60  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.471458 14982 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 61  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.472067 15102 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 62  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.472720 15238 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 63  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.473321 15346 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 64  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.473933 15455 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 65  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.474523 15566 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 66  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.475255 15672 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 67  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.475864 15812 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 68  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.476457 15945 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 69  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
I230318 08:19:05.476944 16063 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 70  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: connection error: desc = "transport: error while dialing: dial tcp 127.0.0.1:40821: connect: connection refused" [code 14/Unavailable]
    context_test.go:1279: initial connection heartbeat failed: grpc: client cluster ID "7dd2d905-beb4-4763-8730-bd741af39120" doesn't match server cluster ID "a27685ad-d0a2-4472-b2de-c3c6f8933c37" [code 2/Unknown]
I230318 08:19:05.484859 15998 rpc/context.go:2249  [rnode=1,raddr=127.0.0.1:40821,class=default,rpc] 71  connection heartbeat loop ended with err: initial connection heartbeat failed: grpc: client cluster ID "7dd2d905-beb4-4763-8730-bd741af39120" doesn't match server cluster ID "a27685ad-d0a2-4472-b2de-c3c6f8933c37" [code 2/Unknown]
--- FAIL: TestHeartbeatHealthTransport (0.09s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

rpc.TestHeartbeatHealthTransport [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9247037?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/9247037?buildTab=artifacts#/) on master @ [0fcc33bc2870961b9387999d8a9fed97fccbb2ae](https://github.com/cockroachdb/cockroach/commits/0fcc33bc2870961b9387999d8a9fed97fccbb2ae):


```
W230325 09:38:45.632785 4695 rpc/context_test.go:1208  [-] 177  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.632827 4695 rpc/context_test.go:1208  [-] 178  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.632864 4695 rpc/context_test.go:1208  [-] 179  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.632903 4695 rpc/context_test.go:1208  [-] 180  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.632938 4695 rpc/context_test.go:1208  [-] 181  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.632969 4695 rpc/context_test.go:1208  [-] 182  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633003 4695 rpc/context_test.go:1208  [-] 183  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633037 4695 rpc/context_test.go:1208  [-] 184  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633072 4695 rpc/context_test.go:1208  [-] 185  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633114 4695 rpc/context_test.go:1208  [-] 186  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633147 4695 rpc/context_test.go:1208  [-] 187  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633185 4695 rpc/context_test.go:1208  [-] 188  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633219 4695 rpc/context_test.go:1208  [-] 189  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633252 4695 rpc/context_test.go:1208  [-] 190  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633291 4695 rpc/context_test.go:1208  [-] 191  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633325 4695 rpc/context_test.go:1208  [-] 192  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633361 4695 rpc/context_test.go:1208  [-] 193  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633394 4695 rpc/context_test.go:1208  [-] 194  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633427 4695 rpc/context_test.go:1208  [-] 195  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633461 4695 rpc/context_test.go:1208  [-] 196  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633498 4695 rpc/context_test.go:1208  [-] 197  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633531 4695 rpc/context_test.go:1208  [-] 198  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633569 4695 rpc/context_test.go:1208  [-] 199  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633602 4695 rpc/context_test.go:1208  [-] 200  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633632 4695 rpc/context_test.go:1208  [-] 201  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633664 4695 rpc/context_test.go:1208  [-] 202  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633702 4695 rpc/context_test.go:1208  [-] 203  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633738 4695 rpc/context_test.go:1208  [-] 204  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633816 4695 rpc/context_test.go:1208  [-] 205  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633865 4695 rpc/context_test.go:1208  [-] 206  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633902 4695 rpc/context_test.go:1208  [-] 207  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.633956 4695 rpc/context_test.go:1208  [-] 208  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634005 4695 rpc/context_test.go:1208  [-] 209  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634059 4695 rpc/context_test.go:1208  [-] 210  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634095 4695 rpc/context_test.go:1208  [-] 211  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634149 4695 rpc/context_test.go:1208  [-] 212  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634205 4695 rpc/context_test.go:1208  [-] 213  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634265 4695 rpc/context_test.go:1208  [-] 214  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634303 4695 rpc/context_test.go:1208  [-] 215  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634362 4695 rpc/context_test.go:1208  [-] 216  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634423 4695 rpc/context_test.go:1208  [-] 217  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634486 4695 rpc/context_test.go:1208  [-] 218  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634545 4695 rpc/context_test.go:1208  [-] 219  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634603 4695 rpc/context_test.go:1208  [-] 220  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634660 4695 rpc/context_test.go:1208  [-] 221  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
W230325 09:38:45.634716 4695 rpc/context_test.go:1208  [-] 222  close tcp 127.0.0.1:34611 -> 127.0.0.1:49756: use of closed network connection
I230325 09:38:45.642603 5730 2@rpc/context.go:2564  [rnode=1,raddr=127.0.0.1:34611,class=default,rpc] 223  connection is now ready
    context_test.go:1262: close tcp 127.0.0.1:34611->127.0.0.1:49756: use of closed network connection
I230325 09:38:45.642928 5730 rpc/context.go:2302  [rnode=1,raddr=127.0.0.1:34611,class=default,rpc] 224  connection heartbeat loop ended with err: grpc: context canceled [code 1/Canceled]
--- FAIL: TestHeartbeatHealthTransport (0.05s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestHeartbeatHealthTransport.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/rpc/context_test.go
+++ b/pkg/rpc/context_test.go
@@ -1176,7 +1176,9 @@ func TestHeartbeatHealthTransport(t *testing.T) {
 		defer mu.Unlock()
 		n := len(mu.conns)
 		for i := n - 1; i >= 0; i-- {
-			if err := mu.conns[i].Close(); err != nil {
+			// This can spuriously return ErrClosed since the listener is closed
+			// before us.
+			if err := mu.conns[i].Close(); err != nil && !errors.Is(err, net.ErrClosed) {
 				return 0, err
 			}
 			mu.conns = mu.conns[:i]
@@ -1272,10 +1274,16 @@ func TestHeartbeatHealthTransport(t *testing.T) {
 		return nil
 	})
 
+	// TODO(baptist): Better understand when this happens. It appears we can get
+	// spurious connections to other tests on a stress run. This has been
+	// happening for a while, but only comes out rarely when this package is
+	// stressed. This test is very aggressive since it is calling GRPCDialNode in
+	// a busy loop for 50ms.
+	expected := "doesn't match server cluster ID"
 	// Should stay unhealthy despite reconnection attempts.
 	for then := timeutil.Now(); timeutil.Since(then) < 50*clientCtx.Config.RPCHeartbeatTimeout; {
 		err := clientCtx.TestingConnHealth(remoteAddr, serverNodeID)
-		if !isUnhealthy(err) {
+		if !isUnhealthy(err) && !testutils.IsError(err, expected) {
 			t.Fatal(err)
 		}
 	}

